### PR TITLE
Fix timeout for blocking read repair when there is a pending node

### PR DESCRIPTION
**Pull request by Runtian**
```
Fix timeout for blocking read repair when there is a pending node

```

The [Cassandra Jira]([CASSANDRA-19120](https://issues.apache.org/jira/browse/CASSANDRA-19120))

### CODE REVIEW

File: src/java/org/apache/cassandra/locator/ReplicaPlans.java [hunk 1]
```diff
@@ -463,7 +455,7 @@ E select(ConsistencyLevel consistencyLevel, L liveAndDown, L live)
                     int add = consistencyLevel.blockForWrite(liveAndDown.replicationStrategy(), liveAndDown.pending()) - contacts.size();
                     if (add > 0)
                     {
-                        E all = consistencyLevel == ConsistencyLevel.LOCAL_QUORUM ? live.all().sorted(replicasInDcComparator) : live.all();
+                        E all = consistencyLevel.isDatacenterLocal() ? live.all().filter(InOurDcTester.replicas()) : live.all();
```
Comment (smiklosovic) [inline src/java/org/apache/cassandra/locator/ReplicaPlans.java:458]: @Runtian 

Blake said "So after the removal of read_repair_chance/dclocal_read_repair_chance in [CASSANDRA-13910](https://issues.apache.org/jira/browse/CASSANDRA-13910), there should never be a case where a DC local read or read repairs are talking to remote DCs."

So I think this is wrong, no? We should just do `live.all().filter(InOurDcTester.replicas())`

File: src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java [hunk 2]
```diff
@@ -205,7 +199,9 @@ public void maybeSendAdditionalWrites(long timeout, TimeUnit timeoutUnit)
         if (awaitRepairsUntil(timeout + timeoutUnit.convert(mutationsSentTime, TimeUnit.NANOSECONDS), timeoutUnit))
             return;
 
-        EndpointsForToken newCandidates = writePlan.liveUncontacted();
+        EndpointsForToken newCandidates = writePlan.liveUncontacted().filter(
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java]: can you rework this bit as:
`writePlan.consistencyLevel().isDatacenterLocal() ? writePlane.liveUncontacted().filter(InOutDcTester.replicas()) : writePlane.liveUncontacted();` so we're only filtering if we need to?

File: src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java [hunk 3]
```diff
@@ -61,31 +60,32 @@ public class BlockingPartitionRepair
...
     {
         this.key = key;
         this.pendingRepairs = new ConcurrentHashMap<>(repairs);
         this.writePlan = writePlan;
-        this.shouldBlockOn = shouldBlockOn;
+
+        // make sure all the read repair targets are contact of the repair write plan
+        if (!all(repairs.keySet(), (r) -> writePlan.contacts().contains(r)))
+        {
+            throw new AssertionError("All repair targets should be part of contacts of read repair write plan.");
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java]: Can you use `Preconditions.checkState` here, or throw an `IllegalStateException`?

File: src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java [hunk 4]
```diff
@@ -61,31 +60,32 @@ public class BlockingPartitionRepair
...
         {
-            // remote dcs can sometimes get involved in dc-local reads. We want to repair
-            // them if they do, but they shouldn't interfere with blocking the client read.
-            if (!repairs.containsKey(participant) && shouldBlockOn.test(participant.endpoint()))
+            if (!repairs.containsKey(participant))
                 blockFor--;
+
+            // make sure for local consistency, all contacts are local replicas
+            if (writePlan.consistencyLevel().isDatacenterLocal() && !InOurDcTester.replicas().test(participant))
+                throw new AssertionError("Local consistency blocking read repair is trying to contact remote DC node: " + participant.endpoint());
```
Comment (bdeggleston) [inline src/java/org/apache/cassandra/service/reads/repair/BlockingPartitionRepair.java]: Can you use `Preconditions.checkState` here, or throw an `IllegalStateException`?

**Review by bdeggleston — CHANGES_REQUESTED** [review]

**Pull request closed without merging**